pkg/template/controller/metrics.go

```diff
@@ -0,0 +1,71 @@
+package controller
+
+import (
+	"time"
+
+	templateapi "github.com/openshift/origin/pkg/template/apis/template"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	kapi "k8s.io/kubernetes/pkg/api"
+)
+
+func newTemplateInstancesTotal() *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "TemplateInstanceController_TemplateInstances_total",
+			Help: "Counts TemplateInstance objects by condition type and status",
+		},
+		[]string{"type", "status"},
+	)
+}
+
+func newTemplateInstancesWaiting() prometheus.Histogram {
+	return prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "TemplateInstanceController_TemplateInstances_active_waiting_time_seconds",
+			Help:    "Show the distribution of active TemplateInstance waiting times",
+			Buckets: []float64{60, 300, 600, 1200, 3600},
+		},
+	)
+}
+
+func (c *TemplateInstanceController) Describe(ch chan<- *prometheus.Desc) {
+	newTemplateInstancesTotal().Describe(ch)
+	newTemplateInstancesWaiting().Describe(ch)
+}
+
+func (c *TemplateInstanceController) Collect(ch chan<- prometheus.Metric) {
+	now := time.Now()
+
+	templateInstances, err := c.lister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	templateInstancesTotal := newTemplateInstancesTotal()
+	templateInstancesWaiting := newTemplateInstancesWaiting()
+
+	for _, templateInstance := range templateInstances {
+		waiting := true
+
+		templateInstancesTotal.WithLabelValues("", "").Inc()
+
+		for _, cond := range templateInstance.Status.Conditions {
+			templateInstancesTotal.WithLabelValues(string(cond.Type), string(cond.Status)).Inc()
```

Comment:
Heads up, during the build metrics review, @smarterclayton was big on following the code stylings of kube_state_metrics for actually registering the metrics.

See `addCountGauge` and `addTimeGauge` in pkg/build/metrics/prometheus/metrics.go

Reply:
I'm reluctant to follow this coding approach because it duplicates logic that sits in the prometheus client library.  My approach uses the client library logic to populate all the histogram buckets before sending them off.  Although it matters less than the Histogram case, `func (bc *buildCollector) Collect` is basically reimplementing the prometheus client Gauge.Inc() when I don't believe it needs to.